internal/controller/pkg/revision/runtime_provider.go

```diff
@@ -0,0 +1,250 @@
+	svc := manifests.Service()
+	if err := h.client.Apply(ctx, svc); err != nil {
+		return errors.Wrap(err, errApplyProviderService)
+	}
+
+	secClient := manifests.TLSClientSecret()
+	secServer := manifests.TLSServerSecret()
+
+	if secClient == nil || secServer == nil {
+		// We should wait for the provider revision reconciler to set the secret
+		// names before proceeding creating the TLS secrets
+		return nil
+	}
+
+	if err := h.client.Apply(ctx, secClient); err != nil {
+		return errors.Wrap(err, errApplyProviderSecret)
+	}
+	if err := h.client.Apply(ctx, secServer); err != nil {
+		return errors.Wrap(err, errApplyProviderSecret)
+	}
+
+	if err := initializer.NewTLSCertificateGenerator(secClient.Namespace, initializer.RootCACertSecretName,
+		initializer.TLSCertificateGeneratorWithOwner(pr.GetOwnerReferences()),
+		initializer.TLSCertificateGeneratorWithServerSecretName(secServer.GetName(), initializer.DNSNamesForService(svc.Name, svc.Namespace)),
+		initializer.TLSCertificateGeneratorWithClientSecretName(secClient.GetName(), []string{pr.GetName()})).Run(ctx, h.client); err != nil {
+		return errors.Wrapf(err, "cannot generate TLS certificates for %q", pr.GetLabels()[v1.LabelParentPackage])
+	}
+
+	return nil
+}
+
+// Post performs operations meant to happen after establishing objects.
+func (h *ProviderHooks) Post(ctx context.Context, pkg runtime.Object, pr v1.PackageRevisionWithRuntime, manifests ManifestBuilder) error { //nolint:gocyclo // this is just slightly over our limit, i.e. 11 > 10.
+	po, _ := xpkg.TryConvert(pkg, &pkgmetav1.Provider{})
+	providerMeta, ok := po.(*pkgmetav1.Provider)
+	if !ok {
+		return errors.New("not a provider package")
+	}
+	if pr.GetDesiredState() != v1.PackageRevisionActive {
+		return nil
+	}
+
+	sa := manifests.ServiceAccount()
+	if err := h.client.Apply(ctx, sa); err != nil {
+		return errors.Wrap(err, errApplyProviderSA)
+	}
+
+	d := manifests.Deployment(sa.Name, providerDeploymentOverrides(providerMeta, pr)...)
+	if err := h.client.Apply(ctx, d); err != nil {
+		// Note(turkenh): Previously, we were using the provider "meta" name in
+		// the selector of the deployment. However, we changed the selector to
+		// use the package name, derived from "v1.LabelParentPackage" instead.
+		// "meta" name and "v1.LabelParentPackage" differs when the package is
+		// installed with a different name, e.g. "upbound-provider-aws-family"
+		// whereas meta name is "provider-aws-family".
+		// Below is a migration since deployment selectors are immutable and
+		// we cannot change it in-place.
+		// TODO(turkenh): Remove this migration in a future release, i.e. > v1.17 (v1.14  + 3 supported release ).
+		if kerrors.IsInvalid(err) && strings.Contains(strings.ToLower(err.Error()), "field is immutable") {
+			if err = h.client.Delete(ctx, d); err != nil {
+				return errors.Wrap(err, errDeleteProviderDeployment)
+			}
+			return errors.New("provider deployment was immutable, deleted and needs to be recreated")
```

Comment:
Do you think it's pretty safe to delete and recreate the provider deployment? An alternative would be to surface an error asking the operator to do it manually. That way they'd get more control over the timing.

Reply:
Good point. 

I believe it is pretty safe (at least it should be safe) to delete any pod in a k8s cluster at any time, as it could also happen for other reasons (e.g. a node going down). However, it may surface some unrelated issues (e.g. a provider not being resistant to deletes during some operation) or cluster being loaded due to many providers (i.e. plenty of family providers) deleted altogether and need to be recreated simultaneously. 

If we leave this to users, the experience would be:

1. User upgrades to v1.14.
2. All providers and active provider revision will report healthy `false` with an error message asking the user to delete the provider deployment manually. 
3. Providers will still be functional and operating fine until the user does the delete.
4. The user must delete provider deployments to get them back to reporting healthy.

Reply:
A reminder on why I want to change the selectors (even though it is breaking): 

Previously, we were using the package name from `crossplane.yaml` in the package, which requires the package to be parsed, which we no longer [want to do](https://github.com/crossplane/crossplane/pull/4071) unless we have to. However, we could also use `pkg.crossplane.io/package` label on the revision, which is set to the name of the Package resource (e.g. Provider). This way, creating a manifest builder wouldn't require pulling/parsing package. 

This is migration I would like to do eventually, but maybe we can postpone it to the next release (so that we can do it early in the release cycle and get the behavior tested on master for a while) 🤔 For now, I can override those selectors for providers with the package name from provider meta for backward compatibility.

Reply:
Just pushed a commit for this: https://github.com/crossplane/crossplane/pull/4744/commits/a85fd86f50a2e384e1d3f41f44e922a09ef62b85

Reply:
While testing, I noticed that we will anyway cause restarting provider pods since we made some changes in the TLS secret names, etc. So, the outcome wouldn't be different. 

Still, this could be done separately later and doesn't have to be part of this (already big) PR.